#' Calculates SNP-level FST values for use in setupGSEArun.R
#'
#' @param genotype_file (char) path to file with SNP genotype data (PLINK format).
#' @param fam_file (char) path to PLINK population coded fam file.
#' @param output_folder (char) path to directory to store output FST files generated by PLINK.
#' @param out_file (char) path to final SNP FST file used for GSEA input.
#'
#' @return none
#' @export
#'

calcFST <- function(genotype_file, fam_file, output_folder, out_file) {

  # Set up PLINK to calculate MAF between case/control populations
  str1 <- sprintf("PLINK --bed %s.bed --bim %s.bim --fam %s", genotype_file, genotype_file, fam_file)
  str2 <- sprintf("--fst case-control --allow-no-sex")
  str3 <- sprintf("--out %s", sprintf("%s/%s", output_folder, basename(genotype_file)))
  cmd  <- sprintf("%s %s %s", str1, str2, str3)
  system(cmd)

  # Read in generated plink.fst file
  fst_cc  <- fread(sprintf("%s/%s.fst", output_folder, basename(genotype_file)),
		h=TRUE, data.table=FALSE)
  fst_input <- subset(fst_cc, select=c(SNP, FST))
  fst_input[is.na(fst_input)] <- 0

  # NOTE: header of column 2 is labelled "CHI2" for purposes of reading into
  # calculate_gsea.pl. This script explicitly accepts two header labels,
  # 'P' or 'CHI2' typically from GWAS. The column still contains SNP FST values.
  cat(sprintf("* Writing out FST file to %s.\n", out_file))
  write.table(fst_input, file=out_file,
		col.names=c("Marker", "CHI2"), row=FALSE, sep="\t", quote=FALSE)
}
